src/main/java/edu/cornell/kfs/coa/batch/service/impl/CopyLegacyAccountAttachmentsServiceImpl.java

```diff
@@ -0,0 +1,202 @@
+import org.kuali.kfs.datadictionary.legacy.DataDictionaryService;
+import org.kuali.kfs.kim.api.identity.PersonService;
+import org.kuali.kfs.kim.impl.identity.Person;
+import org.kuali.kfs.krad.bo.Attachment;
+import org.kuali.kfs.krad.bo.Note;
+import org.kuali.kfs.krad.service.AttachmentService;
+import org.kuali.kfs.krad.service.NoteService;
+import org.kuali.kfs.krad.util.KRADConstants;
+import org.kuali.kfs.krad.util.NoteType;
+import org.kuali.kfs.krad.util.ObjectUtils;
+import org.kuali.kfs.sys.KFSConstants;
+import org.springframework.core.io.buffer.DataBuffer;
+import org.springframework.transaction.annotation.Transactional;
+
+import edu.cornell.kfs.coa.CuCOAKeyConstants;
+import edu.cornell.kfs.coa.batch.CopyLegacyAccountAttachmentsStep;
+import edu.cornell.kfs.coa.batch.CuCoaBatchParameterConstants;
+import edu.cornell.kfs.coa.batch.businessobject.LegacyAccountAttachment;
+import edu.cornell.kfs.coa.batch.dataaccess.CopyLegacyAccountAttachmentsDao;
+import edu.cornell.kfs.coa.batch.service.CopyLegacyAccountAttachmentsService;
+import edu.cornell.kfs.coa.batch.service.DownloadLegacyAccountAttachmentsService;
+import edu.cornell.kfs.sys.util.CuAttachmentUtils;
+
+public class CopyLegacyAccountAttachmentsServiceImpl implements CopyLegacyAccountAttachmentsService {
+
+    private static final Logger LOG = LogManager.getLogger();
+
+    private CopyLegacyAccountAttachmentsDao copyLegacyAccountAttachmentsDao;
+    private DownloadLegacyAccountAttachmentsService downloadLegacyAccountAttachmentsService;
+    private AccountService accountService;
+    private PersonService personService;
+    private NoteService noteService;
+    private AttachmentService attachmentService;
+    private DataDictionaryService dataDictionaryService;
+    private ParameterService parameterService;
+    private ConfigurationService configurationService;
+
+    @Transactional
+    @Override
+    public boolean copyLegacyAccountAttachmentsToKfs() {
+        final int batchSize = getParameterValueAsInteger(CuCoaBatchParameterConstants.ATTACHMENT_BATCH_SIZE);
+        final int maxRetries = getParameterValueAsInteger(CuCoaBatchParameterConstants.MAX_DOWNLOAD_RETRIES);
+        final List<LegacyAccountAttachment> attachmentsToCopy = copyLegacyAccountAttachmentsDao
+                .getLegacyAccountAttachmentsToCopy(batchSize, maxRetries);
+        if (attachmentsToCopy.size() == 0) {
+            LOG.info("copyLegacyAccountAttachmentsToKfs, There are no legacy attachments available for copying");
+            return true;
+        }
+
+        LOG.info("copyLegacyAccountAttachmentsToKfs, Copying {} legacy account attachments into KFS",
+                attachmentsToCopy.size());
+        final List<LegacyAccountAttachment> copiedAttachments = new ArrayList<>(attachmentsToCopy.size());
+        final List<LegacyAccountAttachment> copyFailures = new ArrayList<>();
+        
+        for (final LegacyAccountAttachment attachmentToCopy : attachmentsToCopy) {
```

Comment:
Do you think downloading an attachment, adding a BO note to the account and updating the status of the LegacyAccountAttachment should be in a transactional block?
That would avoid a condition where an account has an attachment added but we don't update the LegacyAccountAttachment status.  That would also prevent subsequent errors preventing updates from saving.

Reply:
The job is currently running in one large transaction, but I'll look into adjusting things so that each account attachment is handled in a new transaction that's not tied to the large one.